arangod/Aql/IResearchViewExecutor.tpp

```diff
@@ -1412,122 +1439,265 @@ bool IResearchViewHeapSortExecutor<ExecutionTraits>::fillBufferInternal(
+  bool gotData = false;
   auto reset = [&] {
-    ++_readerOffset;
-    _currentSegmentPos = 0;
-    _itr.reset();
-    _doc = nullptr;
+    reader.currentSegmentPos = 0;
+    reader.itr.reset();
+    reader.doc = nullptr;
   };
-  for (; _readerOffset < count;) {
-    if (!_itr) {
-      if (!this->_indexReadBuffer.empty()) {
-        // We may not reset the iterator and continue with the next reader if
-        // we still have documents in the buffer, as the cid changes with each
-        // reader.
-        break;
-      }
-
-      if (!resetIterator()) {
+  while (reader.atMost) {
+    if (!reader.itr) {
+      if (!resetIterator(reader)) {
         reset();
-        continue;
+        return gotData;
       }
 
       // segment is constant until the next resetIterator().
       // save it to don't have to look it up every time.
       if constexpr (Base::isMaterialized) {
-        _segment = &this->_reader->segment(_readerOffset);
+        reader.segment = &this->_reader->segment(reader.readerOffset);
       }
-      this->_isMaterialized = false;
-      this->_indexReadBuffer.reset();
     }
 
     if constexpr (Base::isMaterialized) {
-      TRI_ASSERT(_pkReader.itr);
-      TRI_ASSERT(_pkReader.value);
+      TRI_ASSERT(reader.pkReader.itr);
+      TRI_ASSERT(reader.pkReader.value);
     }
     LocalDocumentId documentId;
-
     // try to read a document PK from iresearch
-    bool const iteratorExhausted = !readPK(documentId);
+    bool const iteratorExhausted = !readPK(documentId, reader);
 
     if (iteratorExhausted) {
       // The iterator is exhausted, we need to continue with the next
       // reader.
       reset();
-      if (gotData) {
-        // Here we have at least one document in _indexReadBuffer, so we may not
-        // add documents from a new reader.
-        break;
-      }
-      continue;
+      return gotData;
     }
     if constexpr (Base::isMaterialized) {
       // The collection must stay the same for all documents in the buffer
-      TRI_ASSERT(_segment == &this->_reader->segment(_readerOffset));
+      TRI_ASSERT(reader.segment ==
+                 &this->_reader->segment(reader.readerOffset));
       if (!documentId.isSet()) {
         // No document read, we cannot write it.
         continue;
       }
     }
+    auto current = bufferIdx.load(std::memory_order_relaxed);
+    if constexpr (parallel) {
+      while (!bufferIdx.compare_exchange_weak(current, current + 1)) {
+      }
+    }
+    auto& viewSegment = this->_reader->segment(reader.readerOffset);
     if constexpr (Base::isLateMaterialized) {
-      this->_indexReadBuffer.pushValue(this->_reader->segment(_readerOffset),
-                                       _doc->value);
+      if constexpr (parallel) {
+        this->_indexReadBuffer.setValue(current, viewSegment,
+                                        reader.doc->value);
+      } else {
+        this->_indexReadBuffer.pushValue(viewSegment, reader.doc->value);
+      }
```

Comment:
```suggestion
      this->_indexReadBuffer.makeValue(current, viewSegment,
                                       reader.doc->value);
```

two `makeValue` overloads
```
(size_t, ...) // was setValue
(PushTag, ...) // was pushValue
```

Reply:
And make `current` as `PushTag` instance?

Reply:
Yes